sdk/rust-examples/audio-event-triangulation/src/main.rs

```diff
@@ -0,0 +1,138 @@
+//! An implementation of triangulation of audio event data
+//!
+//! This demo takes in audio events in the form of a timestamp,
+//! GPS coordinates, and an analog rich audio sample, calculates the
+//! best effort triangulated location, returning the estimated location
+//!
+//! ## Authors
+//!
+//! The Veracruz Development Team.
+//!
+//! ## Copyright
+//!
+//! See the file `LICENSING.markdown` in the Veracruz root directory for licensing
+//! and copyright information.
+
+use anyhow;
+use std::{
+    convert::TryInto,
+    fs,
+    io,
+};
+
+struct AudioEvent {
+    #[allow(dead_code)]
+    timestamp: u32,
+    location: (i32, i32),
+    samples: Vec<i16>,
+}
+
+impl AudioEvent {
+    fn power(&self) -> f64 {
+        // compute as avg per single sample, assumes common bitrate
+        let mut p: f64 = 0.0;
+        for x in self.samples.iter() {
+            p += x.abs() as f64;
+        }
+
+        p / (self.samples.len() as f64)
+    }
+}
+
+fn decode_audio_event(event: &[u8]) -> AudioEvent {
+    AudioEvent{
```

Comment:
If any of the input files aren't stored in a correct format then this will abort at runtime due to the `unwrap()`, right?  Can this be made more robust, e.g. introducing an explicit error type that's serialised to the output file indicating either success or failure (with a reason), rather than just aborting?

Reply:
Can do 👍 This was originally written before the WASI changes landed, and I think error reporting in the examples wasn't great then (or I misunderstood it)

Reply:
Yeah, I think there's some older examples that just crash if their inputs are not as they expect, which is probably not what we want to happen.  Rather, I think all examples should be total, in the sense that they always return a defined value via one of their output files (modulo bugs), even if that value is an error from an explicit success type.

Reply:
I was going to suggest we could probably find a tool that would statically determine if the Rust code is total, so we could check for this at compile time, but then I remembered that for code to be panic free in Rust this:

``` rust
a.pow(2) + b.pow(2) + 1
```

Would need to become this:

``` rust
a.checked_pow(2)?.checked_add(b.checked_pow(2)?)?.checked_add(1)?
```

And now I'm not so sure anymore.

Reply:
Yeah, I think manual auditing of the examples in code review (and going back and rewriting some of them to make them total, in cases where they are not) is fine.